tempest/api/workloadmgr/barbican/test_barbican.py

```diff
@@ -2980,3 +2980,125 @@ def test_9_barbican(self):
+                                        "openstack-setup.conf should have luks volume types like - luks-lvm, luks-ceph, luks etc." \
+                                        "Cannot continue with the test...", tvaultconf.FAIL)
+                reporting.set_test_script_status(tvaultconf.FAIL)
+                raise Exception("No luks volume type found to create encrypted volume.")
+
+            if (unencrypted_vol_type == -1):
+                reporting.add_test_step("Unecrypted volume type is missing." \
+                                        "openstack-setup.conf should have volume types like - lvm, ceph, rbd etc." \
+                                        "Cannot continue with the test...", tvaultconf.FAIL)
+                reporting.set_test_script_status(tvaultconf.FAIL)
+                raise Exception("No volume type like lvm, ceph, rbd etc. found to create unencrypted volume.")
+
+            # Now create unencrypted volume with derived volume type id...
+            volume_id = self.create_volume(
+                volume_type_id=unencrypted_vol_type)
+
+            # Attach volume to vm...
+            self.attach_volume(volume_id, vm_id)
+            LOG.debug("volume id attached : " + str(volume_id))
+
+            # create a workload with encryption status as disabled and try to attach it to vm...
+            try:
+                wid = self.workload_create([vm_id],
+                                           tvaultconf.workload_type_id, encryption=False,
+                                           workload_cleanup=True)
+
+                LOG.debug("Workload ID: " + str(wid))
+                reporting.add_test_step("Unencrypted workload creation with unencrypted volume.",
+                                        tvaultconf.PASS)
+                reporting.set_test_script_status(tvaultconf.PASS)
+
+            except Exception as e:
+                LOG.error("Exception: " + str(e))
+                reporting.add_test_step("Unencrypted Workload creation with unencrypted volume.",
+                                        tvaultconf.FAIL)
+                reporting.set_test_script_status(tvaultconf.FAIL)
+
+
+            # Now create an encrypted volume with derived volume type id...
+            volume_id1 = self.create_volume(
+                volume_type_id=encrypted_vol_type)
+
+            # Attach volume to existing vm instance...
+            self.attach_volume(volume_id1, vm_id)
+            LOG.debug("encrypted volume id attached : " + str(volume_id1))
+
+            total_volumes = self.get_attached_volumes(vm_id)
+            LOG.debug("total number of vols: " + str(total_volumes))
+            if len(total_volumes) == 2:
+                reporting.add_test_step("Attach Encrypted volume to existing instance", tvaultconf.PASS)
+                reporting.set_test_script_status(tvaultconf.PASS)
+
+            # create snapshot of workload created...
+            snapshot_id = self.workload_snapshot(wid, True)
```

Comment:
will this step execute even if self.attach_volume fails to attach encrypted volume? Instead of throwing failure in snapshot phase, we can error out here.

Reply:
self.attach_volume() is not failing. It is able to attach encrypted volume to existing instance. After that when we try to take a full snapshot of that workload, error is occurring. 
It is failing in workload_snapshot() and it is handled in else part. For the same, I made modification in getSnapshotInfo() in base.py to return error_msg to handle error gracefully.